rclrs/src/timer.rs

```diff
@@ -0,0 +1,416 @@
+/// on `rcl_clock_t` and `rcl_context_t` by ensuring that these dependencies are
+/// [dropped after][1] the `rcl_timer_t`.
+///
+/// [1]: <https://doc.rust-lang.org/reference/destructors.html>
+pub struct TimerHandle {
+    rcl_timer: Mutex<rcl_timer_t>,
+    _clock: Clock,
+    _context_handle: Arc<ContextHandle>,
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl TimerHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_timer_t> {
+        self.rcl_timer.lock().unwrap()
+    }
+}
+
+impl Drop for TimerHandle {
+    fn drop(&mut self) {
+        let rcl_timer = self.rcl_timer.get_mut().unwrap();
+        let _lifecycle_lock = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
+        // SAFETY: The entity lifecycle mutex is locked to protect against the risk of
+        // global variables in the rmw implementation being unsafely modified during cleanup.
+        unsafe {
+            rcl_timer_fini(rcl_timer);
+        }
+    }
+}
+
+/// Trait to be implemented by concrete [`Timer`]s.
+pub trait TimerBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &TimerHandle;
+    /// Tries to call the timer and run the associated callback.
+    /// Timers are allowed to modify themselves while being executed.
+    fn execute(&mut self) -> Result<(), RclrsError>;
+}
+
+/// A struct for triggering a callback at regular intervals.
+///
+/// If created via [`Node::create_timer()`][1], calling [`spin_once`][2] or [`spin`][3]
+/// on the timer's node will wait until the configured period of time has passed since
+/// the timer was last called (or since it was created) before triggering the timer's callback.
+///
+/// If created via [`Timer::new`], [`is_ready`][4] must be polled until the timer has
+/// expired, after which [`execute`][5] must be called to trigger the timer's callback.
+/// The timer can also be added to a [`WaitSet`][6] to block until it is ready.
+///
+/// [1]: crate::Node::create_timer
+/// [2]: crate::spin_once
+/// [3]: crate::spin
+/// [4]: crate::Timer::is_ready
+/// [5]: crate::Timer::execute
+/// [6]: crate::WaitSet
+pub struct Timer {
+    callback: Arc<Mutex<dyn FnMut(&mut Timer) + Send + Sync>>,
+    handle: TimerHandle,
+}
+
+impl Timer {
+    /// Creates a new `Timer` with the given period and callback.
+    /// Periods greater than i64::MAX nanoseconds will saturate to i64::MAX.
+    ///
+    /// Note that most of the time [`Node::create_timer`][1] is the better way to make
+    /// a new timer, as that will allow the timer to be triggered by spinning the node.
+    /// Timers created with [`Timer::new`] must be checked and executed by the user.
+    ///
+    /// [1]: crate::Node::create_timer
+    pub fn new<F>(
+        context: &Context,
+        clock: Clock,
+        period: Duration,
+        callback: F,
+    ) -> Result<Self, RclrsError>
+    where
+        F: FnMut(&mut Timer) + 'static + Send + Sync,
```

Comment:
Could we get away with a lifetime guarantee something like the suggested? What is the benefit of a 'static lifetime here?
```suggestion
    pub fn new<'a, F>(
        context: &'a Context,
        clock: Clock,
        period: Duration,
        callback: F,
    ) -> Result<Self, RclrsError>
    where
        F: FnMut(&mut Timer) + 'a+ Send + Sync,
```

Reply:
This one often gets me. Keep in mind the `'static` trait bound is subtly different from the `'static` reference lifetime (as per [rust by example](https://doc.rust-lang.org/rust-by-example/scope/lifetime/static_lifetime.html#trait-bound)). 

By my understanding the practical implication of a `'static` trait bound (as seen here) is that Timer needs to own the callback, since that fulfills the requirement that "the receiver can hold on to the type for as long as they want and it will never become invalid until they drop it". This includes owning anything captured by a closure being used as the callback.

If we wanted to switch to lifetimes it would probably need to be more like:

```rust
    pub fn new<'a, F>(
        context: &Context,
        clock: Clock,
        period: Duration,
        callback: F,
    ) -> Result<Timer<'a>, RclrsError>
    where
        F: FnMut(&mut Timer) + 'a + Send + Sync,
```

Since the callback needs to life at least as long as the Timer class, so they're what need to share the lifetime (the &Context just contains an Arc that can be cloned).

This works and does mean you can pass in a `&mut` to closures used for the callback, but there's a trade-off in that we can no longer store the Timer class inside an Arc.

The trade-off comes from Rusts [subtyping and variance](https://doc.rust-lang.org/nomicon/subtyping.html#subtyping-and-variance) system. I've only got a basic understanding of it, but it seems that storing a value inside an Arc erases the lifetime references (as Arc is invariant over 'a), so the compiler won't allow it.

With that in mind, the benefits of explicitly tracking the lifetime doesn't seem to outweigh the costs. Closures can easily "own" any data they reference just by using `Arc<Mutex>` types, working around an inability to use `&mut`, but there doesn't seem to be a neat workaround to not being able to put `Timer` classes inside `Arc`s.

All of that is to say I think we're best served by sticking with `'static` here.

Reply:
That makes good sense, thanks for digging into it and providing a nice succinct answer!